server/pulp/server/managers/consumer/applicability.py

```diff
@@ -344,7 +345,7 @@ def _get_consumer_profile_map(consumer_ids):
 
         # Get all unit profiles associated with given consumers
         unit_profile_criteria = Criteria(
-            filters={'consumer_id': {'$in': consumer_ids}, 'profile': {'$ne': []}},
+            filters={'consumer_id': {'$in': consumer_ids}},
```

Comment:
Do I understand correctly that, alternatively, you could have kept it here and add the` 'profile': {'$ne': []}` to the applicability retrieval part https://github.com/pulp/pulp/pull/3990/files#diff-497c4c3e2dfb13249190d0776ea94a5eL706?

I believe filtering out the empty profile works well the way you suggested, we just carry them and filter out at the last moment. FWIW, The goal was to get rid of them at early stages.

Thanks for the patch! I'm asking more to check my understanding of the problem. No need to fix this part.

Reply:
@hao-yu , actually, thinking more about it... Right now applicability is calculated for `all_profile_hash` which does not include an empty profile. So your fix wouldn't help to retrieve already calculated applicability. User will need to retrigger applicability calculation. 
If we fix the retrieval part, the all_profile_hash will be calculated correctly and users will get the already calculated applicability for consumers.

Secondly, someone might clean up consumer empty profiles (I saw support folks doing that) and that will make applicability not being retrieved if it's not recalculated for the same issue as we see now.

It seems that it's safer to fix the retrieval part https://github.com/pulp/pulp/pull/3990/files#diff-497c4c3e2dfb13249190d0776ea94a5eL706.

What do you think? Any flaw in my logic?
Let me know if my comment is unclear, the topic is complicated and not easy to follow :/